keps/sig-storage/20190204-kubernetes-csi-release-process.md

```diff
@@ -0,0 +1,277 @@
+   declared for the "master" branch and only changes relevant for that next
+   release are accepted.
+
+1. When all changes targeted for the release are in master, automatic
+   test results are okay and and potentially some more manual tests,
+   maintainers tag a new release directly on the master branch. This
+   can be a `-rc` test release or a normal release.
+
+1. Maintenance releases are prepared by creating a "release-X.Y" branch based on
+   release "vX.Y" and backporting relevant fixes from master. The same
+   prow job as for master also handles the maintenance branches.
+
+
+### Implementation Details
+
+For each component under kubernetes-csi (`external-attacher`,
+`csi-driver-host-path`, `csi-lib-utils`, etc.), these Prow jobs need
+to be defined:
+- `kubernetes-csi-<component>-pr: presubmit job
+- `kubernetes-csi-<component>-build: a postsubmits job that matches against
+  `v*` branches *and* tags (see https://github.com/kubernetes/test-infra/pull/10802#discussion_r248900281)
+
+A periodic job that does regular maintenance tasks (like checking for
+updated dependencies) might be added in the future.
+
+These `kubernetes-csi` Prow job all provide the same environment where
+the component is already checked out in the `GOPATH` at the desired
+revision (PR merged tentatively into a branch or a regular
+branch). This is provided by the
+[podutils](https://github.com/kubernetes/test-infra/blob/master/prow/pod-utilities.md)
+decorators. The base image is the latest
+[kubekins](https://github.com/kubernetes/test-infra/tree/master/images/kubekins-e2e)
+image.
+
+The Prow job transfers control to a `.prow.sh` shell script which must
+be present in a component that is configured to trigger the Prow job.
+
+Each component has its own release configuration (what to build and
+publish) and rules (scripts, makefile). The advantage is that those
+can be branched and tagged together with the component. The version of
+Go to use for building is also part of that configuration. The
+requested version of Go will be installed from https://golang.org/dl/
+if different from what is installed already.
+
+To simplify maintenance and ensure consistency, the common parts can
+be shared via
+[csi-release-tools](https://github.com/kubernetes-csi/csi-release-tools/).
+
+Unit testing is provided by `make test`. Images are pushed with `make
+push`, which already determines image tags automatically. For Prow,
+the image destination still needs to be determined
+(https://github.com/kubernetes/k8s.io/issues/158).
+
+For testing on Kubernetes, a real cluster can be brought up with
+[kubetest](https://github.com/kubernetes/test-infra/tree/master/kubetest). This
+might work with [kind](https://github.com/kubernetes-sigs/kind) and a
+locally built image can be pushed directly into that cluster with
+`docker save image | docker exec -it kind-control-plane docker load
+-` (to be tested).
+
+A shared E2E test could work like this:
+- build one CSI app from source
```

Comment:
what is a CSI app?  Is that a sidecar?  Can you add a section with more specifics on permutations of image versions we want to have?  ie:

1. Every sidecar can have a pull job, where that sidecar is locally built, but all the other sidecars in the hostpath deployment are using the latest stable release.
2. A "stable" job for the hostpath csi driver, which tests only officially released sidecars and just uses the versions checked into the repo
3. A "latest sidecar" job for the hostpath csi driver, which tests canary versions of all the sidecars

Reply:
Yes, with "CSI app" I meant a kubernetes-csi app (sidecar or driver).

The shared job that I described here is this job number one in your list.

The "stable" job can be added. It would be a periodic job that runs that same test for the latest release of the `csi-driver-host-path` repo without rebuilding any image. But that's basically testing the same thing over and over, because all parameters are fixed. Or shall we vary something else, like "always test against latest Kubernetes master"?

The "latest sidecar" job is similar, except that it uses the prebuilt canary images. This job is conceptually problematic, because there is no guarantee that the deployment defined in the `csi-driver-host-path` repo actually works for the canary images. We could mitigate that by having a deployment for the latest stable releases (basically what we have now) and a second deployment for canary (which would be the same most of the time).

Reply:
The stable job could be the hostpath driver tests that's running currently in k/k.

For the latest job, having separate deployments for stable and canary is reasonable.  This is what we do for pd driver testing.

Reply:
I've add a section on periodic jobs (= "stable" and "latest sidecar").